### ISSUE TITLE
Support for switching users without daemonizing when running under systemd

### ISSUE THREAD

**Issue by fmorgner** [umurmur/umurmur#105]
As already described in PR #80, it is best-practice to not daemonize when running as a systemd server. However, we would still need to drop privileges without daemonizing but **ONLY** when running as a systemd service. I currently have no idea what the best solution for this problem would be, so comments and suggestions are very welcome!

**Comment by C4K3** [comment]
We could add another flag to use for whether to switch user, -s (--switch-user or --systemd.)

We could either have -s override -d making it not change user, so systemd users would run -d -s, or we could make them exclusive such that -s works as -d except it also changes users. I'm not sure which makes most sense.

**Comment by fmorgner** [comment]
Proposed fix in #107

**Comment by omor1** [comment]
My current systemd service actually just runs umurmurd directly as a lower-privileged user/group. I should probably change that, as it means that the private key can be read as a non-root user... I am using Let's Encrypt though, so I regenerate the certificate every 2/3 months anyways.

**Comment by fmorgner** [comment]
You might be interested in tracking #91, even though I am fairly sure it won't make it into th 0.2.17 release.

**Comment by fdutheil** [comment]
Hi, 
What is the need implying running umurmur as root (as port >1024)? Having a dedicated system user running it and being the only one able to read the private key seems like a common best practice. Whatever the init solution is, starting it like omor1 directly as a non root user is simple and efficient as it involves no code (KISS) ;)

**Comment by omor1** [comment]
@fdutheil I think the primary reason is that it's generally recommended to keep SSL private keys as readable only by root. But we don't want to keep these privileges, so after reading the key we switch to a lower-privileged user.

This behavior supports the use case of e.g. Let's Encrypt, which stores a current live SSL certificates in a directory readable only by root. Currently I copy them out of that directory and set them owned by the umurmur user, but that is tedious to do every time I need to renew the certificate and is less than ideal.

**Comment by fdutheil** [comment]
Hi and thank you for taking the time to discuss this :)

@omor1: 
> the primary reason is that it's generally recommended to keep SSL private keys as readable only by root.

That's a kinda "twisted" (\*) interpretation of the notion of "private" key. Basically (and from the security point of view) a "private" key only means that nobody else than the legitimate user/owner of the key must be able to read it. 
Root has nothing to do here. Worse, it would imply that every piece of software wanting to do TLS crypto should have root permissions.
 _(\*): I mean no offense, I can't think of another word to describe this idea (english not my mother tongue)._



> This behavior supports the use case of e.g. Let's Encrypt, which stores a current live SSL certificates in a directory readable only by root. Currently I copy them out of that directory and set them owned by the umurmur user, but that is tedious to do every time I need to renew the certificate and is less than ideal.

Oh, I see how the confusion arised. Let me try to convince you ;) 
1. The LE client you use (I bet it's the official/vanilla one) needs to be root. It's not a security feature to secure the access to your certificates, but a handy feature provided by the LE client that needs to automatically edit and modify configuration files on your systems. Many other clients (like dehydrated, the one I use) don't do configuration modification thus don't require you to be root to use them. And they still set proper filesystem permissions to secure access to your certs/keys.
2. LE mantra is: renew your certificates often, and do it automatically to avoid it being a pain. Which means if you have to request a certificate and then do a cp and a chown on your certs every 90days, damn, write a 3 lines shell script and enjoy well deserved serenity ;)


So, to sum up:
- the way we manage our certificates must not imply any complex/unsure/unnecessary + highly privileged code in the software that use them. Especially if it's only a chown/chmod away  ;)
- private keys only need to be private regarding the system user that use them. Especially in the scenario of a system user that runs a service.
- KISS everywhere and everyone happy :)

**Comment by Nauxuron** [comment]
> private keys only need to be private regarding the system user that use them. Especially in the scenario of a system user that runs a service.

If someone finds an RCE in umurmur, it becomes possible to extract those private keys. The point of having private keys readable by root only is to make sure that not even the service itself can read them after dropping privileges, because you cannot fully trust the umurmur process in case of an RCE. You only consider the case when another service is compromised, but umurmur itself isn't necessarily safe from it either.

> the way we manage our certificates must not imply any complex/unsure/unnecessary + highly privileged code in the software that use them. Especially if it's only a chown/chmod away ;)

Dropping root privileges isn't that complex and I think it only makes umurmur more secure as stated above.

**Comment by omor1** [comment]
@fdutheil you are correct that retaining the key in the memory of the lower-privileged process still exposes it to attack. One solution, albeit at the cost of increasing complexity, is have only the higher-privileged process read and retain the key in memory. All cryptography would need to go through the higher-privileged process and key itself would never be exposed to the lower-privileged code.

**Comment by fdutheil** [comment]
OK, thank you gentlemen for taking the time to detail what you mean to achieve regarding the root owned private keys/user switching mecanism.
